src/pudl/transform/eia930.py

```diff
@@ -0,0 +1,224 @@
+"""Module to perform data cleaning functions on EIA930 data tables."""
+
+import pandas as pd
+from dagster import AssetOut, Output, multi_asset
+
+import pudl
+
+logger = pudl.logging_helpers.get_logger(__name__)
+
+
+@multi_asset(
+    outs={
+        "core_eia930__hourly_balancing_authority_net_generation": AssetOut(
+            io_manager_key="parquet_io_manager"
+        ),
+        "core_eia930__hourly_balancing_authority_demand": AssetOut(
+            io_manager_key="parquet_io_manager"
+        ),
+    },
+    compute_kind="pandas",
+)
+def core_eia930__hourly_balancing_authority_assets(
+    raw_eia930__balance: pd.DataFrame,
+):
+    """Transforms raw_eia923__balance dataframe.
+
+    Extract the net generation information from the balance table and extract fuel type
+    from the net generation columns.
+    """
+    qual_cols = [
+        "report_datetime_local",
+        "report_datetime_utc",
+        "balancing_authority_code_eia",
+        # report_date
+        # eia_region_code,
+        # report_hour_local
+    ]
+    # Select only the columns relevant to the BA net generation table
+    netgen = raw_eia930__balance[
+        qual_cols
+        + list(raw_eia930__balance.filter(like="net_generation"))
+        + list(raw_eia930__balance.filter(like="interchange"))
+    ]
+
+    # Rename columns so that they contain only the energy source and the level of
+    # processing, so it's easy to construct a multi-index to unstack below
+    # Doing these manipulations while the values are in the column names rather than
+    # after they've been turned into a categorical column with millions of entries
+    # is much faster.
+    netgen_renamed = netgen.rename(
+        lambda col: col.removeprefix("net_generation_").removesuffix("_mw"),
+        axis="columns",
+    ).set_index(qual_cols)
+
+    # Prepare a multi-index for the columns so that we can stack cleanly
+    netgen_renamed.columns = pd.MultiIndex.from_tuples(
+        [x.split("_") for x in netgen_renamed.columns], names=["energy_source", None]
+    )
+    netgen_stacked = (
+        netgen_renamed.stack(level=0, future_stack=True)
+        .rename(columns=lambda x: f"net_generation_{x}_mw")
+        .reset_index()
+        .astype({"energy_source": "string"})
+        .astype(
+            {
+                "balancing_authority_code_eia": pd.CategoricalDtype(),
+                "energy_source": pd.CategoricalDtype(),
+            }
+        )
```

Comment:
Setting these to be `category` saves lots of memory usage, but this should be happening automatically now for `energy_source` as it's getting written out to Parquet and back in through the IO Manager.  The BA ID has a FK relationship with the BA code table, so not sure that it gets to be a category (it's not ENUM constrained, but FK constrained).